draft/XXXX-orm/README.md

```diff
@@ -0,0 +1,103 @@
+---
+RFC: XXXX
+Author: Jacob Nesbitt
+Status: draft
+Created: 2020-07-28
+Last-Modified: 2020-07-28
+---
+
+# RFC XXXX: Multinet ORM
+
+This RFC addresses the issue of organizing database operations into classes which interact with each other. This aims to reduce the complexity, while improving the clarity and extendability of our database operations.
+
+## Motivation
+
+Currently, our database operations are structured in a purely functional way. That is, all database operations are done through simple functions, with no use of classes. This becomes problematic for readability, as our current `db.py` file has grown to almost 600 lines long. This presents a risk, as lack of code readability/clarity likely leads to more mistakes/bugs being introduced.
+
+Furthermore, due to the flat use of stateless functions, performance becomes an issue, as the same few functions are called repeatedly by multiple functions, as they share no state. The temporary solution to this was caching, but by introducing an ORM, much of the need for this caching will go away, as this state can be preserved in a class instance.
+
+## Proposals
+
+### Classes
+The following subsections describe the classes that make up the Multinet ORM. Type annotations are included to more accurately describe the connections between the classes. If a type annotation is left out, it is either due to uncertainty, or because that method returns nothing useful.
+
+
+##### User
+```
+# Previously used the term "cookie"
+def get_session() -> str: ...
+def set_session(): ...
+def delete_session(): ...
+
+# Serializes the object to JSON
+def serialize() -> Dict: ...
+
+# Static
+def exists(user: UserInfo) -> bool: ...
+def register(user: UserInfo) -> User: ...
+def from_session(session_id: str) -> User: ...
+```
+
+##### Table
+```
+def rows() -> Iterable[Dict]: ...
+def row_count() -> int: ...
+def keys() -> Iterable[str]: ...
+def edge() -> bool: ...
+```
+
+##### Graph
+```
+def nodes() -> Iterable[Dict]: ...
+def node_tables() -> Iterable[Table]: ...
+def edge_table() -> Table: ...
+```
+
+
+##### Workspace
+The workspace class would contain the following methods:
+```
+
+def table(name: str) -> Table: ...
+def tables() -> Iterable[Table]: ...
+
+def graph(name: str) -> Graph: ...
+def graphs() -> Iterable[Graph]: ...
+
+def metadata() -> Dict: ...
+
+def rename(new_name: str): ...
+def delete(): ...
+
+def create_table(name: str, rows: Iterable[Dict], edge: bool = False) -> Table: ...
+def create_aql_table(name: str, aql: str) -> Table: ...
+def delete_table(name: str): ...
+
+def create_graph(
+  name: str,
+  edge_table: str,
+  from_collections: Iterable[str],
+  to_collections: Iterable[str],
```

Comment:
Are these necessary? We can automatically figure out what values these should have, so are we simply doing that somewhere else, and leaving this method to match up with the Arango-level functions for creating a graph?

Reply:
I stubbed this out based on the [existing function](https://github.com/multinet-app/multinet-server/blob/264089818733eb3bbcad00495a9526f7ccc487ea/multinet/db.py#L474-L500):

But you're right, when we're calling that function, we're just passing in information that it could derive anyway. I'll update the definition.

Reply:
Well, maybe there's good reason to retain this stuff at the DB layer? If we compute it automatically in the API layer or something, we'd still need to pass the info along to ArangoDB.

Reply:
I'm not sure what you mean. Currently, in the [D3 JSON uploader](https://github.com/multinet-app/multinet-server/blob/264089818733eb3bbcad00495a9526f7ccc487ea/multinet/uploaders/d3_json.py#L124-L132), we're calling:

```
properties = util.get_edge_table_properties(workspace, edge_table_name)
```

and then immediately passing the result into `db.create_graph`. We also pass `workspace` and `edge_table_name` into `db.create_graph`, so there's no need to make this call happen outside of the `create_graph` function. That's all I was saying.